pinot-common/src/main/java/org/apache/pinot/common/utils/helix/TableCache.java

```diff
@@ -0,0 +1,189 @@
+  }
+
+  public String getActualColumnName(String tableName, String columnName) {
+    String schemaName = _tableConfigChangeListener._table2SchemaConfigMap.get(tableName.toLowerCase());
+    if (schemaName != null) {
+      return _schemaChangeListener.getColumnName(schemaName, columnName);
+    }
+    return columnName;
+  }
+
+  class TableConfigChangeListener implements IZkChildListener, IZkDataListener {
+
+    Map<String, TableConfig> _tableConfigMap = new ConcurrentHashMap<>();
+    Map<String, String> _tableNameMap = new ConcurrentHashMap<>();
+    Map<String, String> _table2SchemaConfigMap = new ConcurrentHashMap<>();
+
+    public synchronized void refresh() {
+      try {
+        //always subscribe first before reading, so that we dont miss any changes
+        _propertyStore.subscribeChildChanges(PROPERTYSTORE_TABLE_CONFIGS_PREFIX, _tableConfigChangeListener);
+        _propertyStore.subscribeDataChanges(PROPERTYSTORE_TABLE_CONFIGS_PREFIX, _tableConfigChangeListener);
+        List<ZNRecord> children =
+            _propertyStore.getChildren(PROPERTYSTORE_TABLE_CONFIGS_PREFIX, null, AccessOption.PERSISTENT);
+        if (children != null) {
+          for (ZNRecord znRecord : children) {
+            try {
+              TableConfig tableConfig = TableConfig.fromZnRecord(znRecord);
+              String tableNameWithType = tableConfig.getTableName();
+              _tableConfigMap.put(tableNameWithType, tableConfig);
+              String rawTableName = TableNameBuilder.extractRawTableName(tableNameWithType);
+              //create case insensitive mapping
+              _tableNameMap.put(tableNameWithType.toLowerCase(), tableNameWithType);
+              _tableNameMap.put(rawTableName.toLowerCase(), rawTableName);
+              //create case insensitive mapping between table name and schemaName
+              _table2SchemaConfigMap.put(tableNameWithType.toLowerCase(), rawTableName);
+              _table2SchemaConfigMap.put(rawTableName.toLowerCase(), rawTableName);
+            } catch (IOException e) {
+              e.printStackTrace();
+              //ignore
+            }
+          }
+        }
+      } catch (Exception e) {
+        e.printStackTrace();
+        //ignore
+      }
+    }
+
+    @Override
+    public void handleChildChange(String s, List<String> list)
+        throws Exception {
+      refresh();
+    }
+
+    @Override
+    public void handleDataChange(String s, Object o)
+        throws Exception {
+      refresh();
+    }
+
+    @Override
+    public void handleDataDeleted(String s)
+        throws Exception {
+      refresh();
+    }
+  }
+
+  class SchemaChangeListener implements IZkChildListener, IZkDataListener {
+    Map<String, Schema> _schemaMap = new ConcurrentHashMap<>();
+    Map<String, Map<String, String>> _schemaColumnMap = new ConcurrentHashMap<>();
+
+    public synchronized void refresh() {
+      try {
```

Comment:
So if the listener is invoked for a change to a particular table, we are going to go over schema of all the tables to update the map? If so, I think we should handle the callback at a table level.

Reply:
It makes the logic more complicated and there is batching where we can get only one callback and multiple tables changed in the same callback. We can instead optimize reading the contents by reading the stats. But given that schema and table configs rarely change. I dint bother optimizing it.